### ISSUE TITLE
Issue when connection is lost - status stays as connected

### ISSUE THREAD

**Issue by wagner-rebello** [shamblett/mqtt_client#299]
I'm using a list of connections to connect to many brokers in my local network (Each Broker is a device).
When i unplug power from one device (to simulate a connection loss) none of my callbacks run and the connection status stays connected for that broker/device.
Using wireshark i can see that ping never got a response.
![Screenshot from 2021-07-15 17-23-03](https://user-images.githubusercontent.com/70769274/125831335-65f600d9-aabd-4c31-b004-c5316f44e070.png)
When i turn on the device again ALL the callbacks run at the same time that the connection is reestablished. (I have a timer that checks connection status, the status turn to disconnected at that moment)

As the numbers in the image below:
![log](https://user-images.githubusercontent.com/70769274/125831909-5d8fd1d1-431e-44c2-a80c-d3d1c1c14745.png)

1. Was a ping that have response. Between **1.** and **2.** the selected ping in wireshark picture happens.
2. is 5 minutes later when i powered on the device again. And ALL the callbacks run at the same time.

Codes for creating the connections:
```dart
        clients[clients.length - 1].mqtt.autoReconnect = false;
        clients[clients.length - 1].mqtt.keepAlivePeriod = KEEPALIVE; //30 segundos de Ping.
        clients[clients.length - 1].mqtt.onDisconnected = () => _onDisconnected(qlEh); //Ao desconectar chama esse metodo.
        clients[clients.length - 1].mqtt.onConnected = () => _onConnected(qlEh); //Ao conectar chama esse metodo.
        clients[clients.length - 1].mqtt.onAutoReconnect = () => _onAutoReconnect(qlEh);
        clients[clients.length - 1].mqtt.onSubscribed = (String cb) => _onSubscribed(cb, qlEh); //Ao inscrever-se com sucesso chama esse metodo.
        clients[clients.length - 1].mqtt.onSubscribeFail = (String cb) => _onSubFail(cb, qlEh);
        clients[clients.length - 1].mqtt.onUnsubscribed = (String cb) => _onUnSubscribed(cb, qlEh);
        clients[clients.length - 1].mqtt.pongCallback = () => _pong(qlEh); //Ao receber ping corretamente chama esse metodo.
        clients[clients.length - 1].mqtt.connectionMessage = connMess;
```
```dart
void _onDisconnected(int qlClient) {
    if (clients[qlClient].mqtt.connectionStatus.disconnectionOrigin == MqttDisconnectionOrigin.solicited) {
      print('($qlClient - ${clients[qlClient].id}) Solicitada desconexão do Broker');
    } else {
      _limpaTopicos(qlClient);
      print('($qlClient - ${clients[qlClient].id}) Conexão caiu');
      //TODO: Rodar broadcast novamente???
    }
  }

void _pong(int qlClient) {
    print('($qlClient - ${clients[qlClient].id}) MQTT Ping/Pong [${DateTime.now()}]');
}

void _onConnected(int qlClient) {
    print('($qlClient - ${clients[qlClient].id}) Conectado ao Broker');
    //TODO: Roda subs?
    for (var e in subbedTopics) {
      // if (clients[qlClient].mqtt.getSubscriptionsStatus(e) == MqttSubscriptionStatus.doesNotExist) {
      //FIXME: Vai rodar varios subscribes mesmo sem necessidade pois não vai estar inscrito nos topicos que não deveria.
      subscribe(e, 1);
      // }
      // subscribe(e, 1);
    }
  }
```

Turning autoReconnect to true doesn't help.

**Comment by shamblett** [comment]
OK, there are 2 issues here, the client has never taken any action if the broker doesn't respond to its ping request if the connection status is connected, it doesn't disconnect for example. However, just because no one has asked for this thus far it doesn't mean we can't do this, I can add a setting to to keep alive processing that will disconnect if pings are not responded to, that way users can select the behaviour they want.

The second issue, as in why your callbacks are called when you turn your devices on etc. is nothing to do with the client. The client uses the Dart runtime to interface to the network stack, if the runtime says a socket is connected then the client assumes it is, when the runtime says a socket is disconnected then again it takes appropriate action dependant on user settings. If you want to know why/how the Dart runtime decides these things then I suggest you bring this up with the Dart Google guys.

**Comment by wagner-rebello** [comment]
Using wireshark and filtering for TCP i can see that TCP retransmission keep happening until the device is turned on when the connection is reseted.
A setting to disconnect if keep alive don't respond would be cool to solve that problem.

**Comment by shamblett** [comment]
Yes, I'll update the client to allow the user to select whether a broker failing to respond to a ping will cause a disconnect or not and chose the time period they want to wait.

**Comment by shamblett** [comment]
Client updated to add the ping request response fail functionality, package re published at version 9.4.0.

See the API docs and the example mqtt_server_client_keep_alive_disconnect.dart for usage.

**Comment by shamblett** [comment]
Now at 9.4.2

**Comment by wagner-rebello** [comment]
When the connection is lost with the new keepalive disconnection option the MQTT status correctly enter into disconnected but the socket stay opened waiting for a RST packet from server.
Sometimes this leaves to a state that a new connection replaces the variable "client" in file `\connectionhandling\mqtt_client_mqtt_connection_base.dart` leaving the old socket in a unknown state and closing the new created connection with RST packet from the old sock.
This sometimes put the connection in a loop where it creates a new connection overwrite the old client reference with a new reference, receive a RST or FIN from the old connection closes the new and go on.

I think that it's better to close the TCP socket when the keep alive fails with the option "enabled".

**Comment by wagner-rebello** [comment]
After some diggin' I found out that this chunk of code `connectionhandling\server\mqtt_client_synchronous_mqtt_server_connection_handler.dart`
![image](https://user-images.githubusercontent.com/70769274/128076205-bac61140-036c-4b28-a35e-32e1b7533609.png)

Is leaving lost references to the old TCP socks, when the new connection is created the "MqttServerConnection" inheritances creates new references, if old socks listeners capture data (Like RST or FIN) the data is then redirected to the new object references.

The old socks stays de-referenced and got FIN (or RST because of TCP Retransmission) by the server by the lack of keep alive (or TCP Retransmission), the old listener put the state of connection in disconnected, my code call the connection method again that de-reference the actual connection and go on...

**Comment by shamblett** [comment]
I hear what you are saying, the old socket events are somehow redirected to the closures(onDone, onError etc.) of the new socket. But what can the client actually do here? Dart is a GC based language, only when GC runs are unused objects destroyed, even doing client == null probably won't force a GC at that point. All the client can do is call methods on the old socket to destroy it somehow. What do you mean by ' with the option "enabled".

**Comment by wagner-rebello** [comment]
"What do you mean by ' with the option "enabled".
Set the property ("Option") `disconnectOnNoResponsePeriod` >= 1

**Comment by shamblett** [comment]
Ok, where is this in the Dart API? The [SocketOption](https://api.dart.dev/stable/2.13.4/dart-io/SocketOption-class.html) class only has tcpNoDelay, apologies if I'm missing something here.

**Comment by wagner-rebello** [comment]
Examples of the problem using wireshark.
![PingFail](https://user-images.githubusercontent.com/70769274/128177621-d9c2fba7-5521-4998-af5f-604f703c602e.png)
As you can see in the line Nº 287 there was a "ping req" (port 48420) from the smartphone without response (I intentionally removed my device (broker) from power). onDisconnect callback runs without problem informing my code that MQTT connection dropped.
After 3 seconds my code try another connection (Nº 323)(device still offline and flutter sent retransmissions for the SYN packet).

When my device turned on again it ACKed the SYN packet and the TCP connection opened (48423 port).

After some time the old socket tryed another retransmission from the lost ping request (Nº 1461 from below image) (port 48420, first connection)
![RstFirst](https://user-images.githubusercontent.com/70769274/128178327-2e8fd9d0-3ec4-46e4-9d51-6350d6927fb3.png)

As the method _disconnect() from file `\connectionhandling\mqtt_client_mqtt_connection_base.dart` never got called when there is a ping response fail, the old sock still exists (port 48420), including the old listeners.
The device RSTed the connection (Nº 1464) because it was powered off 'cause the connection don't exist, that RST call the onDisconnect callback and set the connection as disconnected also Ping Req from the new connection do not happen anymore (Set to 15s) when my device stays more than 15s without ping req it ends the connection as expected and put the smartphone in this eternal loop.

**Comment by wagner-rebello** [comment]
> 
> 
> "What do you mean by ' with the option "enabled".
> Set the property ("Option") `disconnectOnNoResponsePeriod` >= 1

This is the new option that you created to "disconnect" when there is no ping respones after some time.

**Comment by shamblett** [comment]
BTW issue #302 shows the disconnectOnNoResponsePeriod  opytion working for the use case the OP had, so it does work in some, hopefully most cases.

**Comment by wagner-rebello** [comment]
> 
> 
> Ah OK, its not a socket option, so back to my original point, how do you suggest the client 'disposes' of old references to its sockets? I can't think of a way to do this.

Maybe some chunk of code that can be called on `internalConnect` method that verify existance of connection variable before re-instantiation and some public method that destroy the socks.

I'm creating a clean project to test how the TCP listener behave after destroying sockets.

**Comment by shamblett** [comment]
Yes, [destroy](https://api.dart.dev/stable/2.13.4/dart-io/Socket/destroy.html) may be an option here, depending on how it works, we dont  want destroy to call one of the closures like onDone etc. Your work here is interesting, please post back your results, thanks.

**Comment by wagner-rebello** [comment]
After the tests i can confirm.
Calling destroy on the old sock (client variable in your lib) stop all callbacks from listen.
onData, onError listeners from the old sock are destroyed in conjunction.
But the onDone is called when destroying the old client.

**Comment by shamblett** [comment]
Yes I remember now, I actually use this mechanism in the browser ws comnnection -
```
...
 closeEvents = client.onClose.listen((e) {
        MqttLogger.log(
            'MqttBrowserWsConnection::connect - websocket is closed');
        closeEvents.cancel();
        errorEvents.cancel();
        return completer.complete(MqttClientConnectionStatus());
      });
      errorEvents = client.onError.listen((e) {
        MqttLogger.log(
            'MqttBrowserWsConnection::connect - websocket has erred');
        closeEvents.cancel();
        errorEvents.cancel();
        return completer.complete(MqttClientConnectionStatus());
      });
...
```
Why it never occurred to me to apply this to the rest of the connection handlers I don't know.

I'll update the package to add this functionality.

**Comment by shamblett** [comment]
OK, I've updated the client with the fix above, I've not re published this yet as we may find more problems here.

If you would point your pubspec.yaml to the git repo and set the branch to 'issue299-1' and carry on with your testing that would be great.

**Comment by shamblett** [comment]
I've found a small bug when using disconnect on no ping response with auto reconnect and fixed it, if you are currently testing with this could you please do a pub upgrade, thanks.

**Comment by wagner-rebello** [comment]
Sure, i'l test.

**Comment by wagner-rebello** [comment]
It keeps entering in loop when the connection is lost by keep alive.

I attached some wireshark packet capture.

[PacketCapture.zip](https://github.com/shamblett/mqtt_client/files/6954762/PacketCapture.zip)

I created a fork in my personal account and did some changes on 9.4.2, in my fork after receiving the RST packets from the OLD connection the new stays opened.

https://github.com/ZantsuRocks/mqtt_client/commit/8db8e3bc0866d2d0b3de4ee2f20cf5fb4e7cc589

I didn't really understand the structure of your code, so I didn't create a pull request. My code is not 100% implemented either, but it serves as a basis perhaps for a fix.

**Comment by shamblett** [comment]
Ok, could you just confirm when you do a pub upgrade that the branch it prints is 'issue-299-1' with a commit of 'a42065814a8adf08e8b49b9443c1b06649b1df63'.

**Comment by wagner-rebello** [comment]
> 
> 
> Ok, could you just confirm when you do a pub upgrade that the branch it prints is 'issue-299-1' with a commit of 'a42065814a8adf08e8b49b9443c1b06649b1df63'.

Yes
![image](https://user-images.githubusercontent.com/70769274/128722958-7bdd2bde-e682-4340-b9c4-82c5e88bc4e0.png)

**Comment by wagner-rebello** [comment]
I'm trying to replicate the issue again using your branch to tell you the exact steps that create the problem.
But now the problem isn't happening again.

**Comment by wagner-rebello** [comment]
Gotcha.
![image](https://user-images.githubusercontent.com/70769274/128728635-370f99ed-f2bb-4771-9485-b7e49a13ade6.png)

The problem is with the timing between the `ping req` and the `ping resp`

Ping response delayed something about 120ms to arrive and the `connection` died for your handlers.
This created a duplicate connection and the old still open.

**Comment by shamblett** [comment]
OK, I'm not really following you, what do you mean by ' the connection died for your handlers'? What is causing this, do you have a log?

When you say 'This created a duplicate connection and the old still open.', we now cancel the listener and destroy the socket so how does this happen?

I'll put some more logging around the destroy method to see exactly when it is called.

**Comment by wagner-rebello** [comment]
Put some log inside `_disconnect()` in the `MqttConnectionBase` class.
You'll see that when keep alive fails (There is no Ping Response to a Ping Request) that method never runs.
At least for me this is happening.

I'm put some log in `MqttConnectionBase` class in the method `_disconnect()` (line 80)
https://gist.github.com/wagner-rebello/e906a6af8b9aded8ac3346907173ff9e

This is what happens to my log (the red arrow is the onDisconnect callback)
![image](https://user-images.githubusercontent.com/70769274/128750819-44b2244e-35fa-4e51-a841-d40ed0d5f39c.png)

As you can see the log on `MqttConnectionBase` never runs leaving the sock listener and connection "open" but turning your status (`MqttClientConnectionStatus`) disconnected.

As the connection may still be open TCP Retransmission keep ocurring the old listeners still exists too, wich kills the new connection.

**Comment by wagner-rebello** [comment]
Maybe this log helps more
![image](https://user-images.githubusercontent.com/70769274/128752245-f7723630-42a3-4ba5-9f4f-3bf181d434b8.png)

- The Red Arrow is when i turned of the power from the broker device. (onDisconnect callback from ping without response)

- Blue Arrow is the new connection (onConnect callback after calling again connect as the actual status (MqttClientConnectionStatus) is now MqttConnectionState.disconnected).

- Green Arrow is the old socket getting a TCP RST (TCP keeps sending Ping Request retransmissions until it get a FIN or a RST from the broker).

That Green Arrow put the MqttClientConnectionStatus again in MqttConnectionState.disconnected even that the new socket still exists, and this is where the reconnection loop go on forever.

**Comment by wagner-rebello** [comment]
Pull Request #308 fixes the connection loop for me.
The old socket stay alive when there is a ping fail but the listeners references at least will be destroyed.

**Comment by shamblett** [comment]
Thanks for this, merged the change, please pub upgrade and re test, thanks.

**Comment by shamblett** [comment]
I've added a small update, hopefully it should now close the socket before we re connect, please pub upgrade and see what you get now.

**Comment by wagner-rebello** [comment]
Sure, i'll test.

**Comment by wagner-rebello** [comment]
Changes were tested by 3 people together with me and we didn't find any more problems.

**Comment by shamblett** [comment]
Thanks, I've merged the branch, finished my own testing and re published the package at version 9.5.0.

Thanks for the testing you have done here, I know how time consuming this can be, this helps you of course in your use case but the fixes identified in this issue help all the users which is great.

I'll close this issue here, its got rather lengthy, if you find anything else please raise a separate issue, we can always refer to this one if needed.
